### PR TITLE
[CLI] [FR] Add new CLI command for generating Dockerfile (issue #6532)

### LINKED ISSUE #6532: [FR] create new cli command for generating dockerfile
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

We have ability to create an docker image for ml model, stored and tracked in mlflow with command
`mlflow models build-docker <...> ` 
This command has several arguments\flags that can be specified, full list of which can be found in the [documentation](https://mlflow.org/docs/latest/cli.html#mlflow-models-build-docker) 

However, I find this approach rather opinionated in regard of building docker image. I find it more flexible to be able to extract said dockerfile and build this image using standard docker tools.

I propose a solution to overcome those difficulties without affecting any currently available functionality: **new cli command to generate `dockerfile` only**, which can be manipulated afterwards

### Motivation

> #### What is the use case for this feature?

For example, it can be useful in several scenarios:

- Building docker image on a remote host, even allowing us to use some tools like `buildx` to do so  (oftentimes useful with ci/cd pipelines)
- Changing resulting dockerfile, changing base image (useful when dealing with troublesome installments that have docker image of themselves, such as CUDA)
- Setting up cache policies for apt packages (keeping in mind considerate image size, it can be pretty useful coupled with remote host docker build)

> #### Why is this use case valuable to support for MLflow users in general?

Most of the scenarios are the ones, we are experiencing currently, introducing mlflow to our ml development, so I do believe they are applicable to the variety of users.

I believe, this approach will ease many processes related to integrating mlflow models or mlflow itself in ci/cd pipelines/cycles.

> #### Why is this use case valuable to support for your project(s) or organization?

We, at our organization restructuring our approaches regarding development of ML services. We find mlflow of great value regarding this process. It allows us to concentrate on ML model for ML team and on domain features for product teams, since it can automate "wrapping" our ml model in RESTful service. However we struggle sometimes with our currently existing ci/cd pipelines which were created with mindset of somewhat total control over things like dockerfile, which we must overcome sometimes, since it's easier for us to adapt to existing pipelines rather than changing ci/cd approaches we have

> #### Why is it currently difficult to achieve this use case?

Since mlflow calls docker implicitly it is difficult to pass arguments (such as `-H` for remote build). Python is not a primary language for my organization so it would be very troublesome to fork mlflow, since we coudn't possibly provide adequate support for it.
We've found some ways around those issues (such as setting `DOCKER_HOST` environment variable), but we are struggling right now with "post-build-docker" steps, such as installment of CUDA toolkit. We could possibly use ml model, built with mlflow, as a base image and manipulate over it (to install some software, for example) but this will add a major step to our ci/cd pipelines and will require us to maintain it

Ability to generate dockerfile would ease out those cases since we could change it afterwards and will allow us to reduce complexity of the steps necessary to deploy ml service

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [X] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by anuarkaliyev23**
## Related Issues/PRs

Close #6532 

## What changes are proposed in this pull request?

This pull request proposes changes in CLI commands: it adds new `generate-dockerfile` command for `models`. It the outputs dockerfile with model directory in directory specified by CLI parameter. Resulting Dockerfile upon building and running behaves exactly like `models build-docker` image

example:
```bash
mlflow models generate-dockerfile -m <artifact> -d <name of the output directory>
```

To achieve this many functions of `build_docker` were reused and slightly refactored

## How is this patch tested?

I tested locally both `generate-dockerfile` and `build-docker` commands with seemingly expected behavior. Python is not my primary language, so I would be glad to hear any advise regarding writing adequate tests
- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?
I am not entirely sure. I can see that `generate-dockerfile` cli command is present in `--help`, but I am not sure if any other documentation should be fixed for this matter

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

[CLI] [FR] Adding new CLI command for generating Dockerfile (issue #6532)
    
Implemented cli command to generate dockerfile identical to the one in `build-docker` command

### Is this a user-facing change?

- [] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add new CLI command for generating dockerfiles for model tracked and stored with mlflow

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/cli.py
Comment (harupy) [inline mlflow/models/cli.py:38]: @anuarkaliyev23 Thanks for the PR! Can you revert unrelated changes?

Comment (anuarkaliyev23) [reply mlflow/models/cli.py]: Yes, I'm sorry, my IDE seems to change indents and I didn't notice.
Fixed

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:174]:
```suggestion
    This dockerfile defines an image that is equivalent to the one produced by ``mlflow models build-docker``
```

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:178]: Can we check `backend.can_build_image()` and throw if the backend cannot build an image, which indicates that we cannot build a dockerfile for it? Can we add a similar check to https://github.com/mlflow/mlflow/blob/9ea22cdf4d18204e6b92585a028d40341667ac6e/mlflow/models/cli.py#L209?

File: mlflow/pyfunc/backend.py
Comment (dbczumar) [inline mlflow/pyfunc/backend.py:319]: While we're at it, can we add the `build_image` and `generate_dockerfile` base methods to the FlavorBackend base class (https://github.com/mlflow/mlflow/blob/9ea22cdf4d18204e6b92585a028d40341667ac6e/mlflow/models/flavor_backend.py#L4) with a default implementation of `raise NotImplementedError`?

**Review by dbczumar — COMMENTED** [review]
@anuarkaliyev23 This is looking great! Just left a few small comments - we should be all set to merge after they're addressed. Thank you for your contribution!

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:163]: Can we accept a path rather than a directory, e.g. `/tmp/my/custom/path/Dockerfile` and default to `Dockerfile` so that the default behavior produces a dockerfile in the current working directory?

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:168]: Can we add a test for this command? I think that https://github.com/mlflow/mlflow/blob/master/tests/models/test_cli.py is a good suite to add to

File: mlflow/models/cli.py
Comment (harupy) [reply mlflow/models/cli.py:163]: +1 for using a path. Can we validate the specified Dockerfile path doesn't exist?

File: mlflow/pyfunc/backend.py
Comment (harupy) [inline mlflow/pyfunc/backend.py:348]: Can we revert this change?

File: mlflow/models/docker_utils.py
Comment (harupy) [inline mlflow/models/docker_utils.py:155]:
```suggestion
def _generate_dockerfile_content(
```

Can we rename this function `_generate_dockerfile_content`?

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py:163]: I started with a path, but encountered problems with `model_dir`. At first, I thought, that function will generate a `Dockerfile` only, but since `model_dir` used to copy model into image --- it should be present also.
This led me to believe that even if I pass a desired path to `Dockerfile` it will create a `model_dir ` in a parent directory, which seems to me as a little unexpected\not obvious, which can result in confusion for user --- since it only accepts dockerfile path, user can assume that only dockerfile is needed to build an image and can move it somewhere else *without moving `model_dir`*. 
Maybe I'm a little too cautious here, but folder seems as not ideal, but at least an obvious option

As for validating if a path doesn't exist --- function ensures creation of a path with `os.makedirs`, which allows us to pass any path

I am open to discussions and suggestions --- maybe I am just being too cautious here?

File: mlflow/pyfunc/backend.py
Comment (anuarkaliyev23) [reply mlflow/pyfunc/backend.py]: Done

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py]: Added in both cases :+1:

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py:168]: I will add test once we can find consensus on directory parameter in function, since it can potentially change a function hence a test.
At the moment I'm just not sure if we might need to change a function to accept a path to Dockerfile instead of directory :sweat_smile: , so until you approve decision to leave directory as an argument or decide to change a function's aruments, I will hold on a test for a little bit :sweat_smile:

File: mlflow/models/cli.py
Comment (dbczumar) [reply mlflow/models/cli.py:163]: @anuarkaliyev23 This is a great point. I think that producing a directory is the best option given that wrinkle. Can we rename the default directory to `mlflow-dockerfile` and update the helpstring to indicate that the directory contains a `Dockerfile` and also the resolved model, if `model-uri` is specified?

File: mlflow/models/cli.py
Comment (dbczumar) [reply mlflow/models/cli.py:168]: Responded here: https://github.com/mlflow/mlflow/pull/6591/files#r956774884. I think that a directory output makes sense :)

Comment (anuarkaliyev23) [reply mlflow/models/cli.py:168]: Yep, @dbczumar, sorry for the delay --- we had country holidays up until Wednesday :)
I'm gonna try and update pull request today to address all remaining comments and issues

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:431]:
```suggestion
    output_directory_name = pyfunc_generate_dockerfile(model_uri, extra_args=extra_args)
```

Is this line correct? I think we should call `pyfunc_generate_dockerfile` here.

File: tests/helper_functions.py
Comment (harupy) [inline tests/helper_functions.py:89]:
```suggestion
def pyfunc_generate_dockerfile(output_directory, model_uri=None, extra_args=None):
```

Can we add `output_directory` and pass a temporary directory created by pytest to this argument?

File: tests/helper_functions.py
Comment (harupy) [inline tests/helper_functions.py:110]:
```suggestion
    subprocess.run(cmd, check=True)
```

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:436]:
```suggestion
    output_directory = Path(output_directory_name)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
    assert output_directory.joinpath("Dockerfile").stat().st_size != 0
```

nit

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py]: I have written a test to ensure that `model_dir` and `Dockerfile` are generated and `Dockerfile` is not empty
However, I am somewhat struggling with test launching --- I am running `dev/run-python-tests` but it fails in 3 cases for unrelated reasons (not installed `kubernetes`, `google.cloud` and `pyarrow` modules)
When I tried to purposefully fail tests they are also didn't appear as failed once I ran the script.

Could it be related to issue https://github.com/mlflow/mlflow/issues/1096?

File: tests/models/test_cli.py
Comment (anuarkaliyev23) [reply tests/models/test_cli.py]: Yep, you are correct, my bad :sweat:

File: tests/helper_functions.py
Comment (anuarkaliyev23) [reply tests/helper_functions.py]: Yep, it will be more convenient to do so
Done

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:419]:
```suggestion
def test_generate_dockerfile(sk_model, enable_mlserver, tmp_path):
```

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:439]:
```suggestion
    pyfunc_generate_dockerfile(tmp_path, model_uri, extra_args=extra_args)
    output_directory = Path(temp_dir)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
    assert output_directory.joinpath("Dockerfile").stat().st_size != 0
```

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py]: @harupy @dbczumar  Should we add `kubernetes`, `google-cloud-storage` and `pyarrow` to `requirements/test_requirements.txt` to ensure valid launch of `dev/run-python-tests`?
Should it be a separate issue?

File: mlflow/models/cli.py
Comment (harupy) [inline mlflow/models/cli.py:163]: As @dbczumar suggested, can we change the default value to `mlflow-dockerfile`?

Comment (anuarkaliyev23) [reply mlflow/models/cli.py]: I'm sorry, it might've gotten lost somewhere in stash, I should've checked before resolving

File: tests/models/test_cli.py
Comment (anuarkaliyev23) [reply tests/models/test_cli.py]: @harupy 
I haven't found any fixtures for `tmp_path`
is it some default fixture in pyfunc?
I'm sorry if it is a stupid question -- I'm not experienced in Python at all

Comment (harupy) [reply tests/models/test_cli.py]: > is it some default fixture in pyfunc?

Yes, it's provided by pytest.

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:433]: Looks like `temp_dir` here is undefined.

**Comment by anuarkaliyev23** [comment]
@harupy @dbczumar Should we add kubernetes, google-cloud-storage and pyarrow to requirements/test_requirements.txt to ensure valid launch of dev/run-python-tests?
Should it be a separate issue?

Duplicated as a general comment for easier discussion

**Comment by harupy** [comment]
I think `pip install -e .[extra]` should install `kubernetes`, `google-cloud-storage` and `pyarrow`.

https://github.com/mlflow/mlflow/blob/c9f64728f93331a4b1d70d685c17694fb2ac45fc/setup.py#L130

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:433]:
```suggestion
```

Nit: we can remove this line. `tmp_path` is a `Path` instance.

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:42]:
```suggestion
    PROTOBUF_REQUIREMENT,
    pyfunc_generate_dockerfile,
```

to pass the lint check.

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:437]:
```suggestion
    output_directory = tmp_path.joinpath("output_directory")
    pyfunc_generate_dockerfile(output_directory, model_uri, extra_args=extra_args)
    assert output_directory.is_dir()
    assert output_directory.joinpath("Dockerfile").exists()
    assert output_directory.joinpath("model_dir").is_dir()
    # Assert file is not empty
```

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:6]:
```suggestion
```

Let's remove this. `Path` is no longer used.

File: tests/models/test_cli.py
Comment (harupy) [inline tests/models/test_cli.py:5]:
```suggestion
```

**Review by harupy — APPROVED** [review]
Looks good to me once the lint errors in https://github.com/mlflow/mlflow/runs/8149563761?check_suite_focus=true are fixed :)

**Comment by anuarkaliyev23** [comment]
@harupy 
> Looks good to me once the lint errors in https://github.com/mlflow/mlflow/runs/8149563761?check_suite_focus=true are fixed :)

yep, also spotted one typo in tests

`pylint` seem to take on formatting differently than `black` in case of, for example, line length
Should I address line lengths in pylint if they are seem to be ok in `black` eyes?

**Comment by harupy** [comment]
> Should I address line lengths in pylint errors if they are seem to be ok in black eyes?

Yes. Unfortunately, black doesn't help you format too-long lines.

**Comment by anuarkaliyev23** [comment]
@harupy 
I believe I fixed lint errors
In the previous run there were some failed tests, but they seem to be unrelated to the feature
Tests for CLI were successful

File: mlflow/pyfunc/backend.py
Comment (harupy) [inline mlflow/pyfunc/backend.py:327]:
```suggestion
        custom_setup_steps = copy_model_into_container(output_path)
```

Can we remove `if-else`?

**Comment by anuarkaliyev23** [comment]
@dbczumar 
Can you please review these changes? :)

**Comment by WallEnd** [comment]
@dbczumar, please help us review this PR. We are so waiting for these changes.

**Comment by anuarkaliyev23** [comment]
@harupy 
If @dbczumar is not available could we assign another member to review changes prior to merge?

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:178]:
```suggestion
    Generates a directory with Dockerfile whose default entrypoint serves an MLflow model at port
    8080 using the python_function flavor. The generated Dockerfile is written to the specified output
    directory, along with the model (if specified). This Dockerfile defines an image that is
    equivalent to the one produced by ``mlflow models build-docker``
```

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:164]:
```suggestion
    help="Output directory where the generated Dockerfile is stored",
```

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:179]: Can we add a small `logger.info("Generating Dockerfile)` message here? If a model_uri is specified, can we include it in the message?

File: mlflow/models/cli.py
Comment (dbczumar) [inline mlflow/models/cli.py:189]: After generation completes successfully, can we add a small `logger.info("Generated Dockerfile in directory <directory_path>")` message here?

**Review by dbczumar — APPROVED** [review]
LGTM once two small cosmetic comments are addressed. Thanks @anuarkaliyev23 !

File: mlflow/models/cli.py
Comment (anuarkaliyev23) [reply mlflow/models/cli.py:178]: done

File: mlflow/models/cli.py
Comment (harupy) [inline mlflow/models/cli.py:182]:
```suggestion
        _logger.info("Generating Dockerfile")
```

Do we need `for unspecified model`?

Comment (anuarkaliyev23) [reply mlflow/models/cli.py:182]: @dbczumar asked "Can we add a small logger.info("Generating Dockerfile) message here? If a model_uri is specified, can we include it in the message?"

Maybe I took too far "if model_uri is specified" part :sweat_smile: 

I think we can just add "Generating dockerfile for model <model_uri>" if it's OK to you?
Than it just will be `None` if model_uri was not specified

Comment (dbczumar) [reply mlflow/models/cli.py]: @anuarkaliyev23 Your latest changes look great! Thanks!

**Pull request merged**
